isthmus/src/main/java/io/substrait/isthmus/expression/ExpressionRexConverter.java

```diff
@@ -0,0 +1,159 @@
+import org.apache.calcite.rex.RexInputRef;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.sql.SqlOperator;
+import org.apache.calcite.util.TimeString;
+import org.apache.calcite.util.TimestampString;
+
+public class ExpressionRexConverter extends AbstractExpressionVisitor<RexNode, RuntimeException> {
+  private final RelDataTypeFactory typeFactory;
+  private final RexBuilder rexBuilder;
+  private final ScalarFunctionConverter scalarFunctionConverter;
+
+  public ExpressionRexConverter(
+      RelDataTypeFactory typeFactory, ScalarFunctionConverter scalarFunctionConverter) {
+    this.typeFactory = typeFactory;
+    this.rexBuilder = new RexBuilder(typeFactory);
+    this.scalarFunctionConverter = scalarFunctionConverter;
+  }
+
+  @Override
+  public RexNode visit(Expression.NullLiteral expr) throws RuntimeException {
+    return null;
+  }
+
+  @Override
+  public RexNode visit(Expression.BoolLiteral expr) throws RuntimeException {
+    return rexBuilder.makeLiteral(expr.value());
+  }
+
+  @Override
+  public RexNode visit(Expression.I8Literal expr) throws RuntimeException {
+    return rexBuilder.makeLiteral(expr.value(), TypeConverter.convert(typeFactory, expr.getType()));
+  }
+
+  @Override
+  public RexNode visit(Expression.I16Literal expr) throws RuntimeException {
+    return rexBuilder.makeLiteral(expr.value(), TypeConverter.convert(typeFactory, expr.getType()));
+  }
+
+  @Override
+  public RexNode visit(Expression.I32Literal expr) throws RuntimeException {
+    return rexBuilder.makeLiteral(expr.value(), TypeConverter.convert(typeFactory, expr.getType()));
+  }
+
+  @Override
+  public RexNode visit(Expression.I64Literal expr) throws RuntimeException {
+    return rexBuilder.makeLiteral(expr.value(), TypeConverter.convert(typeFactory, expr.getType()));
+  }
+
+  @Override
+  public RexNode visit(Expression.FP32Literal expr) throws RuntimeException {
+    return rexBuilder.makeLiteral(expr.value(), TypeConverter.convert(typeFactory, expr.getType()));
+  }
+
+  @Override
+  public RexNode visit(Expression.FP64Literal expr) throws RuntimeException {
+    return rexBuilder.makeLiteral(expr.value(), TypeConverter.convert(typeFactory, expr.getType()));
+  }
+
+  @Override
+  public RexNode visit(Expression.StrLiteral expr) throws RuntimeException {
+    return rexBuilder.makeLiteral(expr.value(), TypeConverter.convert(typeFactory, expr.getType()));
+  }
+
+  @Override
+  public RexNode visit(Expression.BinaryLiteral expr) throws RuntimeException {
+    // Calcite RexLiteral only takes ByteString
+    return rexBuilder.makeLiteral(
+        new ByteString(expr.value().toByteArray()),
+        TypeConverter.convert(typeFactory, expr.getType()));
+  }
+
+  @Override
+  public RexNode visit(Expression.TimeLiteral expr) throws RuntimeException {
+    // Expression.TimeLiteral is Micros, while RexLiteral assumes Milliseconds for time type
+    int milliS = (int) (TimeUnit.MICROSECONDS.toMillis(expr.value()));
```

Comment:
Are you sure this is right? I'm thinking that millis should be on a day boundary followed by nanos for anything within a day. E.g. `TimeUnit.DAYS.toMillis(TimeUnit.MICROSECONDS.toDays(expr.value()))` instead of what you have.

Reply:
You are right. There is a bug.  We need a `second` boundary followed by nanos for any fractional second part.

```
    long microSec = expr.value();
    long seconds = TimeUnit.MICROSECONDS.toSeconds(microSec);
    int fracSecondsInNano =
        (int) (TimeUnit.MICROSECONDS.toNanos(microSec) - TimeUnit.SECONDS.toNanos(seconds));
    TimeString timeString =
        TimeString.fromMillisOfDay((int) TimeUnit.SECONDS.toMillis(seconds))
            .withNanos(fracSecondsInNano);
```

Previously, I used 'miliSeconds' as boundary. But turns out that Calcit's TimeString will remove the fractional milliseconds part, if we call `withNanos`.

I added two new UT `tTimeWithMicroSecond`, `tTimeWithNanoSecond`. The second is just to verify Calcite's time literal truncates to microseconds precision, the same precision as Substrait.

1. https://github.com/apache/calcite/blob/main/core/src/main/java/org/apache/calcite/util/TimeString.java#L87-L93

Reply:
Also, I fixed the similar bug in TimeStampLiteral, changing to use seconds boundary. And add new UT testcase for such coverage.